### ISSUE TITLE
Decorator support for pint units

### ISSUE THREAD

**Issue by pheuer** [PlasmaPy/PlasmaPy#1762]
### Feature description

A decorator for functions with dimensional input that automatically converts input and output using various unit packages.

### Motivation

Python's unit handling ecosystem is unfortunately fragmented. In addition to [astropy.units](https://docs.astropy.org/en/stable/units/index.html) (used internally in PlasmaPy) some others include: 

- [pint](https://pint.readthedocs.io/en/stable/) 
- [unyt](https://unyt.readthedocs.io/en/latest/usage.html#working-with-code-that-uses-astropy-units)

For context: 
- astropy -> ~ 4k forks on GitHub
- pint -> ~400 forks
- unyt -> ~40 forks 

If an existing package or code base uses a different unit package, it can be clumsy to integrate PlasmaPy functionality. This potentially limits the use cases for PlasmaPy. 

### Implementation strategy

Create a decorator that inspects arguments and keywords for instances of dimensional quantities from other packages (particularly `pint`). If found, convert those arguments to `astropy.unit.Quantity` instances. If output is dimensional, return it using the same unit package as the input was provided in.

**Comment by namurphy** [comment]
> The fragmented landscape of python unit libraries is potentially a big issue for creating a harmonious software ecosystem.

💯

Yes, the fragmentation is a HUGE problem with the broader scientific Python ecosystem.  Our motivation for going with `astropy.units` was interoperability with the astropythoniverse and the heliopythoniverse, which both heavily depend on Astropy.  Related is that (last I heard) xarray support for units is starting with pint, though I gathered from a pre-pandemic conversation that their goal was to implement it in a way that would allow it to be extended to other units packages. There has been some work within NumPy to generalize dtypes and make the implementation of dtypes less monolithic, which has been a barrier to improving & consistentifying units functionality.

My main thought is that we would want to work in communication and collaboration with the various units packages, in particular because we don't to take a fragmented approach in enabling interoperability. In that sense, I'm thinking that it'd be ideal if this functionality would be made upstream of PlasmaPy, unless we had a compelling or urgent use case.

Also, I just realized that `unyt` is spelled with a `yt` because it was an offshoot of the `yt` package. Only took me three years to realize this!

**Comment by namurphy** [comment]
This also reminded me a talk at SciPy 2017ish on the [problems with the unit ecosystem](https://www.youtube.com/watch?v=qCo9bkT9sow&list=PLYx7XA2nY5GfdAFycPLBdUDOUtdQIVoMf&index=70). Looking back at the suggested solutions near the end of the talk, it made me realize that there really has been a lot of progress in improving units packages in the last five years...even though there still are major issues like the one you brought up.

**Comment by StanczakDominik** [comment]
Yeah, as far as I know our strategy for this has been "let's use astropy.units for now and boy I sure hope those numpy dtype improvements come soon enough!". There's definite progress on that count, but it's still on the way.
